File: internal/domain/entities/prompts.go
package entities

import "time"

type PromptCategory string

const (
	PromptCategoryPostGen     PromptCategory = "post_gen"
	PromptCategoryPageGen     PromptCategory = "page_gen"
	PromptCategoryLinkSuggest PromptCategory = "link_suggest"
	PromptCategoryLinkApply   PromptCategory = "link_apply"
	PromptCategorySitemapGen  PromptCategory = "sitemap_gen"
)

// ContextFieldType defines the UI control type for a context field
type ContextFieldType string

const (
	ContextFieldTypeCheckbox ContextFieldType = "checkbox"
	ContextFieldTypeSelect   ContextFieldType = "select"
	ContextFieldTypeInput    ContextFieldType = "input"
	ContextFieldTypeNumber   ContextFieldType = "number"
	ContextFieldTypeTextarea ContextFieldType = "textarea"
)

// SelectOption represents an option for select-type context fields
type SelectOption struct {
	Value string `json:"value"`
	Label string `json:"label"`
}

// ContextFieldValue represents the configuration for a single context field
type ContextFieldValue struct {
	Enabled bool   `json:"enabled"`
	Value   string `json:"value,omitempty"`
}

// ContextConfig stores the configuration for context fields
type ContextConfig map[string]ContextFieldValue

// ContextFieldDefinition defines metadata for a context field
type ContextFieldDefinition struct {
	Key          string           `json:"key"`
	Label        string           `json:"label"`
	Description  string           `json:"description"`
	Type         ContextFieldType `json:"type"`
	Options      []SelectOption   `json:"options,omitempty"`
	DefaultValue string           `json:"defaultValue"`
	Required     bool             `json:"required"`
	Categories   []PromptCategory `json:"categories"`
	Group        string           `json:"group,omitempty"`
}

// Prompt represents an AI prompt with instructions and context configuration
type Prompt struct {
	ID            int64
	Name          string
	Category      PromptCategory
	IsBuiltin     bool
	Instructions  string        // User's instructions (becomes System Prompt for AI)
	ContextConfig ContextConfig // JSON config for context fields
	Version       int           // 2 = new format, 1 = legacy format
	// Legacy fields (kept for backward compatibility and migration)
	SystemPrompt string
	UserPrompt   string
	Placeholders []string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

// IsV2 returns true if this prompt uses the new v2 format
func (p *Prompt) IsV2() bool {
	return p.Version >= 2
}

File: frontend/src/models/prompts.ts
import { dto } from "@/wailsjs/wailsjs/go/models";

export type PromptCategory = "post_gen" | "page_gen" | "link_suggest" | "link_apply" | "sitemap_gen";

export const PROMPT_CATEGORIES: Record<PromptCategory, string> = {
    post_gen: "Post Generation",
    page_gen: "Page Generation",
    link_suggest: "Link Suggestions",
    link_apply: "Link Insertion",
    sitemap_gen: "Sitemap Structure",
};

// Context field types for v2 prompts
export type ContextFieldType = "checkbox" | "select" | "input" | "textarea";

export interface SelectOption {
    value: string;
    label: string;
}

export interface ContextFieldValue {
    enabled: boolean;
    value?: string;
}

export type ContextConfig = Record<string, ContextFieldValue>;

export interface ContextFieldDefinition {
    key: string;
    label: string;
    description: string;
    type: ContextFieldType;
    options?: SelectOption[];
    defaultValue: string;
    required: boolean;
    categories: string[];
    group?: string;
}

export interface ContextFieldsResponse {
    fields: ContextFieldDefinition[];
    defaultConfig: ContextConfig;
}

// Prompt interface with v2 support
export interface Prompt {
    id: number;
    name: string;
    category: PromptCategory;
    isBuiltin: boolean;
    version: number;
    // V2 fields
    instructions?: string;
    contextConfig?: ContextConfig;
    // Legacy v1 fields
    systemPrompt?: string;
    userPrompt?: string;
    placeholders?: string[];
    createdAt: string;
    updatedAt: string;
}

export interface PromptCreateInput {
    name: string;
    category: PromptCategory;
    version: number;
    // V2 fields
    instructions?: string;
    contextConfig?: ContextConfig;
    // Legacy v1 fields
    systemPrompt?: string;
    userPrompt?: string;
    placeholders?: string[];
}

export interface PromptUpdateInput extends Partial<PromptCreateInput> {
    id: number;
}

export function mapPrompt(x: dto.Prompt): Prompt {
    return {
        id: x.id,
        name: x.name,
        category: (x.category || "post_gen") as PromptCategory,
        isBuiltin: x.isBuiltin || false,
        version: x.version || 1,
        instructions: x.instructions || "",
        contextConfig: x.contextConfig ? mapContextConfig(x.contextConfig) : undefined,
        systemPrompt: x.systemPrompt || "",
        userPrompt: x.userPrompt || "",
        placeholders: x.placeholders || [],
        createdAt: x.createdAt,
        updatedAt: x.updatedAt,
    };
}

export function mapContextConfig(config: Record<string, dto.ContextFieldValue>): ContextConfig {
    const result: ContextConfig = {};
    for (const [key, value] of Object.entries(config)) {
        result[key] = {
            enabled: value.enabled,
            value: value.value,
        };
    }
    return result;
}

export function mapContextFieldDefinition(def: dto.ContextFieldDefinition): ContextFieldDefinition {
    return {
        key: def.key,
        label: def.label,
        description: def.description,
        type: def.type as ContextFieldType,
        options: def.options?.map(opt => ({
            value: opt.value,
            label: opt.label,
        })),
        defaultValue: def.defaultValue,
        required: def.required,
        categories: def.categories,
        group: def.group,
    };
}

export function mapContextFieldsResponse(resp: dto.ContextFieldsResponse): ContextFieldsResponse {
    return {
        fields: resp.fields?.map(mapContextFieldDefinition) || [],
        defaultConfig: resp.defaultConfig ? mapContextConfig(resp.defaultConfig) : {},
    };
}

// Helper to check if a prompt is v2
export function isV2Prompt(prompt: Prompt): boolean {
    return prompt.version >= 2;
}
